File: restaurants.judge.py
from Judger import Judger
import sys
import functools
import imp


class Tester(Judger): 
    def __init__(self, debug=True):
        if debug:
            super().__init__("restaurants.json")
        else:
            super().__init__(sys.argv[1], debug=False, save=True, clean_after_read=True)

    def run(self, sample):
        Restaurant  = imp.load_source("Restaurant" , "restaurants.sol.py").Restaurant
        Restaurants = imp.load_source("Restaurants", "restaurants.sol.py").Restaurants
        rests = [Restaurant(*i) for i in sample[0]['args'][0]]
        self.resetTime()
        if sample[0]['func'] == "init":
            sol = Restaurants(rests)
        out = []
        for sam in sample[1:]:
            if sam['func'] == "filter":
                out.append(getattr(sol, sam['func'])(*sam['args']))
            elif sam['func'] == "sortN":
                out.append([i.getID() for i in sorted(rests)])
            elif sam['func'] == "sortD":
                out.append([i.getID() for i in sorted(rests, key=functools.cmp_to_key(Restaurant.comparator1))])
        return out

    def compare(self, out, sample):
        for i in range(len(out)):
            if type(out[i]) is not list:
                return False
            if out[i] != sample[i + 1]['answer']:
                print(sample[i + 1], out[i])
                return False
        return True


Tester(debug=True).judge()
# Tester(debug=True).judge()

File: restaurants.judge.java
import com.google.gson.JsonElement;
import com.google.gson.JsonArray;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Collections;
/*
dk openjdk:14-slim javac -cp gson.jar:algs4.jar:. restaurants.judge.java restaurants.sol.java Judger.java
dk openjdk:14-slim java -cp gson.jar:algs4.jar:. Judge
*/

class Judge extends Judger<List<int[]>> {

    public Judge(boolean debug, boolean clean_after_read) {
        super(debug);
        this.file_path = "restaurants.json";
        this.clean_after_read = clean_after_read;
    }

    static class Operation {
        JsonElement answer;
        JsonArray args;
        String func;
        public Operation() {};
    }

    @Override protected boolean compare(List<int[]> out, JsonElement ele) {
        Operation[] ops = this.gson.fromJson(ele, Operation[].class);
        for(int i=1; i<ops.length; ++i) {
            switch(ops[i].func){
                case "filter":
                case "sortD":
                case "sortN":
                    if (!Arrays.equals(this.gson.fromJson(ops[i].answer, int[].class),
                                       out.get(i - 1)))
                        return false;
                    break;
            }
        }
        return true;
    }

    @Override protected List<int[]> run(JsonElement ele) {
        Operation[] ops = this.gson.fromJson(ele, Operation[].class);
        List<int[]>[] restaurants = this.gson.fromJson(ops[0].args, new TypeToken<List<int[]>[]>(){}.getType());
        List<int[]> output = new ArrayList<int[]>();
        List<Restaurant> rests = new ArrayList<Restaurant>(),
                         tmprests;
        Restaurants sol;
        int[] ids;
        for(int[] a: restaurants[0])
            rests.add(new Restaurant(a[0], a[1], a[2], a[3]));

        this.resetTime();
        if (!ops[0].func.equals("sort"))
            sol = new Restaurants(rests);
        else
            sol = new Restaurants(new ArrayList<Restaurant>());
        this.updateTime();

        for (int i=1; i<ops.length; i++) {
            switch(ops[i].func) {
                case "filter":
                    this.resetTime();
                    output.add(
                        sol.filter(ops[i].args.get(0).getAsInt(),
                                   ops[i].args.get(1).getAsInt(),
                                   ops[i].args.get(2).getAsInt())
                    );
                    this.updateTime();
                    break;
                case "sortN":
                    tmprests = new ArrayList<Restaurant>(rests);
                    ids = new int[tmprests.size()];
                    this.resetTime();
                    Collections.sort(tmprests);
                    for(int j=0; j<tmprests.size(); ++j)
                        ids[j] = tmprests.get(j).getID();
                    this.updateTime();
                    output.add(ids);
                    break;
                case "sortD":
                    tmprests = new ArrayList<Restaurant>(rests);
                    ids = new int[tmprests.size()];
                    this.resetTime();
                    Collections.sort(tmprests, new Restaurant.Comparator1());
                    for(int j=0; j<tmprests.size(); ++j)
                        ids[j] = tmprests.get(j).getID();
                    this.updateTime();
                    output.add(ids);
                    break;
            }
        }
        this.resetTime();
        return output;
    }

    @Override protected void debugPrint(List<int[]> out, JsonElement ele) {
        for (int[] a: out)
            System.out.println("<" + Arrays.toString(a));
        System.out.println(">" + gson.toJson(ele));
    };

    public static void main(String []args) {
        // Judge j = new Judge(false, true);
        Judge j = new Judge(true, false);
        j.judge(args);
    }
}
